### PR TITLE
feat: Implement selector functions first, last, min, and max

### PR DESCRIPTION
**Pull request by alamb**
This PR implements the InfluxDB 'selector' functions `first`, `last`, `min`, and `max` as DataFusion aggregate functions, planned for use in implementing read_group (https://github.com/influxdata/influxdb_iox/issues/448), as well as implementing first/last/min/max in read_window_aggregate: https://github.com/influxdata/influxdb_iox/issues/480

This PR implements the approach described in https://github.com/influxdata/influxdb_iox/issues/448#issuecomment-744601824 of computing the aggregate twice and here is how it fits into my plan to complete read group support:

- [x] Implement read_group for `None` aggregates (https://github.com/influxdata/influxdb_iox/pull/538)
- [x] Error if any hints are provided, or if there is a discrepancy between `GroupBy` and `GroupNone` type (https://github.com/influxdata/influxdb_iox/pull/539)
- [x] Implement read_group for `Sum`, `Count` and `Mean` aggregates (https://github.com/influxdata/influxdb_iox/pull/557)
- [x] Implement user defined aggregates for `min`, `max`, `first` and `last`, selector functions (this PR)
- [ ] Implement read_group for `Min`, `Max`, `First` and `Last` using selector functions 
- [ ] Update capabilities response that says IOx can support read_group 

I will be filing some arrow improvements shortly and updating the placeholder text in the PR with them

### CODE REVIEW

File: query/src/func/selectors.rs [hunk 1]
```diff
@@ -0,0 +1,688 @@
...
+//!
+//! Selector functions are different than aggregate functions because
+//! they also return multiple column values rather than a single
+//! scalar. Selector functions return the entire row that was
+//! "selected" from the timeseries (value and time pair).
+//!
+//! Note: At the time of writing, DataFusion aggregate functions have
+//! no way to handle aggregates that produce multiple columns.
+//!
+//! TODO: file an arrow ticket to support this behavior
```
Comment (alamb) [inline query/src/func/selectors.rs:14]:
```suggestion
```

**Review by alamb — COMMENTED** [review]
Follow up items are cataloged as : https://github.com/influxdata/influxdb_iox/issues/567 and https://github.com/influxdata/influxdb_iox/issues/568

**Review by carols10cents — APPROVED** [review]
Seems ok, this is pretty complicated! Looks like a good base to profile from though. Having to iterate to get the index whenever you have to update is unfortunate :(

**Comment by alamb** [comment]
@carols10cents  

> Seems ok, this is pretty complicated! Looks like a good base to profile from though. Having to iterate to get the index whenever you have to update is unfortunate :(

I agree basically on all points. This was the simplest I could come up with -- and I did try basically all of yesterday to get this working with generics. 

Oh well, hopefully it will get somewhat simpler with some contributions upstream to arrow

**Comment by alamb** [comment]
https://github.com/apache/arrow/pull/8946 is the PR to move the min / max array kernels to arrow: https://github.com/apache/arrow/pull/8946

**Comment by alamb** [comment]
An update here -- @pauldix  brought to my attention that selector functions may not work the way I thought they did (namely by selecting a single row, but they may return multiple rows). I need to do some more investigation to confirm this implements what it is supposed to be doing before I merge it

**Comment by alamb** [comment]
I did some additional investigation today about the expected output of `read_group` with selector functions. In the [write up](https://docs.google.com/document/d/1UoCov4_NEhJHTxQftn5-YlzfhCiP5hgqDq9ikbRRa3g/edit#heading=h.67cybq49xj0z) I conclude that the functions in this PR do the right thing

**Pull request merged**